In [1]:
import pandas as pd
import statsmodels.formula.api as smf # 回归分析

# 身高与体重的例子

有五位学生，他们的身高分别是152、158、165、172、180厘米；体重分别是43、48、54、63、80公斤。以身高作为自变量、体重作为因变量做回归分析并做显著性检验。

In [2]:
df = pd.DataFrame(
{"hight" : [152, 158, 165, 172, 180],
"weight" : [43, 48, 54, 63, 80]
})

In [6]:
df_model = smf.ols("weight ~ hight",data=df).fit() 
#ols:OLS回归
#"weight ~ hight" ：回归方程表达式，因变量 ~ 自变量
# .fit : fit the model

df_model.summary()

e:\installation\Anaconda\lib\site-packages\statsmodels\stats\stattools.py:74: ValueWarning: omni_normtest is not valid with less than 8 observations; 5 samples were given.
  warn("omni_normtest is not valid with less than 8 observations; %i "


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 weight   R-squared:                       0.954
Model:                            OLS   Adj. R-squared:                  0.939
Method:                 Least Squares   F-statistic:                     62.11
Date:                Wed, 30 Nov 2022   Prob (F-statistic):            0.00426
Time:                        00:41:33   Log-Likelihood:                -12.238
No. Observations:                   5   AIC:                             28.48
Df Residuals:                       3   BIC:                             27.69
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept   -154.8070     26.999     -5.734      0.011    -240.731     -68.883
hight          1.2842      0.163      7.881      0.004       0.766       1.803
==============================================================================
Omnibus:                          nan   Durbin-Watson:                   1.572
Prob(Omnibus):                    nan   Jarque-Bera (JB):                0.592
Skew:                           0.072   Prob(JB):                        0.744
Kurtosis:                       1.320   Cond. No.                     2.77e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.77e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Dep. Variable:	weight	 因变量
R-squared:	0.954  R方
Model:	OLS	       OLS回归
Adj. R-squared:	0.939 调整后的R方
F-statistic:	62.11  F值
Prob (F-statistic):	0.00426 F检验的p值（判断模型是否有统计学意义）



coef	std err	t	P>|t|	[0.025	0.975]
Intercept（截距）	-154.8070	26.999	-5.734	0.011	-240.731	-68.883
回归系数：hight	1.2842	0.163	7.881	0.004	0.766	1.803


# 馅饼销量

## 多个自变量

In [8]:
df_pie = pd.DataFrame(
    {
        "weeks":[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
        "sales":[350, 460, 350, 430, 350, 380, 430, 470, 450, 490, 340, 300, 440, 450, 300],
        "price":[5.5, 7.5, 8, 8, 6.8, 7.5, 4.5, 6.4, 7, 5, 7.2, 7.9, 5.9, 5, 7],
        "advertisement":[3.3, 3.3, 3, 4.5, 3, 4, 3, 3.7, 3.5, 4, 3.5, 3.2, 4, 3.5, 2.7]
    })


In [9]:
pie_model = smf.ols("sales ~ price + advertisement",data=df_pie).fit()

#多个自变量，自变量之间之间用“+”
pie_model.summary()

e:\installation\Anaconda\lib\site-packages\scipy\stats\stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=15
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.521
Model:                            OLS   Adj. R-squared:                  0.442
Method:                 Least Squares   F-statistic:                     6.539
Date:                Thu, 17 Nov 2022   Prob (F-statistic):             0.0120
Time:                        00:14:24   Log-Likelihood:                -77.510
No. Observations:                  15   AIC:                             161.0
Df Residuals:                      12   BIC:                             163.1
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
Intercept       306.5262    114.254      2.683      0.020      57.588     555.464
price           -24.9751     10.832     -2.306      0.040     -48.576      -1.374
advertisement    74.1310     25.967      2.855      0.014      17.553     130.709
==============================================================================
Omnibus:                        1.505   Durbin-Watson:                   1.683
Prob(Omnibus):                  0.471   Jarque-Bera (JB):                0.937
Skew:                           0.595   Prob(JB):                        0.626
Kurtosis:                       2.709   Cond. No.                         72.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

## 自变量出现类别变量

In [18]:
df_pie2 = pd.DataFrame(
    {
        "weeks":[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
        "sales":[350, 460, 350, 430, 350, 380, 430, 470, 450, 490, 340, 300, 440, 450, 300],
        "price":[5.5, 7.5, 8, 8, 6.8, 7.5, 4.5, 6.4, 7, 5, 7.2, 7.9, 5.9, 5, 7],
        "holiday":[0]*12 + [1]*3
    })


In [19]:
df_pie2.holiday = df_pie2.holiday.astype("category")

In [20]:
pie2_model = smf.ols("sales ~ price + holiday",data=df_pie2).fit()

pie2_model.summary()

e:\installation\Anaconda\lib\site-packages\scipy\stats\_stats_py.py:1772: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=15
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  sales   R-squared:                       0.220
Model:                            OLS   Adj. R-squared:                  0.091
Method:                 Least Squares   F-statistic:                     1.697
Date:                Thu, 01 Dec 2022   Prob (F-statistic):              0.224
Time:                        01:46:05   Log-Likelihood:                -81.170
No. Observations:                  15   AIC:                             168.3
Df Residuals:                      12   BIC:                             170.5
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept      579.7795     99.249      5.842      0.000     363.535     796.024
holiday[T.1]   -24.7830     40.804     -0.607      0.555    -113.688      64.122
price          -26.5357     14.420     -1.840      0.091     -57.954       4.883
==============================================================================
Omnibus:                        5.237   Durbin-Watson:                   2.211
Prob(Omnibus):                  0.073   Jarque-Bera (JB):                1.435
Skew:                          -0.096   Prob(JB):                        0.488
Kurtosis:                       1.497   Cond. No.                         44.0
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""